Update design of FontSizePicker when withSlider is set

noisysocks opened the pull request:
## What?
Updates the design and behaviour of `FontSizePicker` when `withSlider` is set to match [the design](https://github.com/WordPress/gutenberg/issues/34345#issue-981044454).

- The slider will now only appear when setting a custom font size value.
- The slider will appear inline with the custom font size field.
- Users can now edit `em` and `rem` values using the slider.

## Why?
Part of an effort to make the Typography panel match the design. See https://github.com/WordPress/gutenberg/issues/34345#issuecomment-1217390408.

## How?


## Testing Instructions
1. Open the site editor.
2. Go to Global Styles → Typography → Text.
3. Select _Set custom size_.
4. Play with setting different custom font sizes using the text field and slider.

## Screenshots or screencast <!-- if applicable -->
**Global styles:**

| Before | After |
| - | - | 
| <img width="298" alt="Screen Shot 2022-09-30 at 15 34 57" src="https://user-images.githubusercontent.com/612155/193198337-73d65cbb-3c88-4a13-b225-fdb9e4e92055.png"> | <img width="295" alt="Screen Shot 2022-09-30 at 15 35 25" src="https://user-images.githubusercontent.com/612155/193198375-dbf65910-b464-4866-8606-3d01d2d70b92.png"> |

**Storybook:**

| Before | After |
| - | - | 
| <img width="340" alt="Screen Shot 2022-09-30 at 15 41 24" src="https://user-images.githubusercontent.com/612155/193198452-23d1bd7a-97c0-40a8-b52c-1f910e20b966.png"> | <img width="330" alt="Screen Shot 2022-09-30 at 15 40 52" src="https://user-images.githubusercontent.com/612155/193198444-e3691948-884d-465a-8a5f-c7ceac2000ea.png"> |

**Storybook with _Reset_ button:**

This isn't used anywhere in Gutenberg (we could probably deprecate `withReset`) but I ensured it looks reasonable as this is the default.

| Before | After |
| - | - | 
| <img width="340" alt="Screen Shot 2022-09-30 at 15 41 24" src="https://user-images.githubusercontent.com/612155/193198649-3f105029-e84d-4986-8185-56dec66779c1.png"> | <img width="341" alt="Screen Shot 2022-09-30 at 15 43 28" src="https://user-images.githubusercontent.com/612155/193198662-2e436dd3-bd09-4cfd-bb38-eced4784a4cd.png"> |

**Video:**

https://user-images.githubusercontent.com/612155/193198882-05f6f042-5b7c-4166-823b-727db01ca602.mp4

noisysocks committed:
Update design of FontSizePicker when withSlider is true

noisysocks committed:
splitNumberAndUnitFromSize -> parseNumberAndUnitFromSize

noisysocks commented on packages/edit-site/src/components/global-styles/style.scss:
```diff
@@ -19,6 +19,7 @@
 	margin-bottom: $grid-unit-20;
 	background: $gray-100;
 	border-radius: $radius-block-ui;
+	overflow: hidden;
```
This stops the `Aa` text breaking out of the font preview container when the font size is really large.

noisysocks commented on packages/components/src/font-size-picker/utils.ts:
```diff
@@ -44,24 +44,20 @@ const FONT_SIZES_ALIASES = [
-		isFinite( Number( numericValue ) )
-	) {
-		return [ numericValue, unit ];
+export function parseNumberAndUnitFromSize(
```
Not really necessary for this PR but I thought I'd update this function to:

- accept `undefined`, since that's a valid input to `FontSizePicker`;
- be more strict about what is returned, i.e.
  - specify a return type;
  - return a `number` instead of a `string` containing a number;
- use a regular regex with capturing groups instead of relying on `/g` behaviour;
- not use "value" in the name since that can be confused with the `value` prop in `FontSizePicker`.

ciampo replied:
`UnitControl` has got a bunch of utils function that potentially achieve a very similar (if not the same) result: https://github.com/WordPress/gutenberg/blob/1326a46cf10ad54e897d1f3c9aaa0e385aba896a/packages/components/src/unit-control/utils.ts

Maybe we could use those function directly? It would remove redundant code AND make sure that we're more consistent on the way we parse unit and values. 

Also happy if we want to leave this task separate and work on it in the future.

noisysocks replied:
Oh fantastic! That lets us delete this altogether 😀

noisysocks commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -169,6 +151,15 @@ const UnforwardedFontSizePicker = (
+	const [ , firstFontSizeUnit ] = parseNumberAndUnitFromSize(
+		fontSizes[ 0 ]?.size
+	);
+	const hasUnits = !! valueUnit || !! firstFontSizeUnit;
```
Also not really necessary for this PR but was irking me. Previously `hasUnits` was calculated by checking the type of `value` and the first option's `size`. If either was a string then we'd allow units. This isn't 100% correct though since technically a font size can be a string containing a number with no unit, i.e. `"30"`. This new logic uses `parseNumberAndUnitFromSize` which is a little more tedious but correct.

noisysocks replied:
I was wrong about this. See https://github.com/WordPress/gutenberg/pull/44598#issuecomment-1277002334.

noisysocks commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -201,141 +192,145 @@ const UnforwardedFontSizePicker = (
+							}: {
+								selectedItem: FontSizeSelectOption;
+							} ) => {
+								if ( selectedItem.size === undefined ) {
```
I changed all of the `onChange` callbacks to explicitly handle the `undefined` case because:

- I think it's easier to understand, and;
- if `hasUnits` is `false` then you could get into a situation where `onChange` is called with `NaN` because `Number( undefined ) === NaN`.

noisysocks commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -201,141 +192,145 @@ const UnforwardedFontSizePicker = (
+											onChange?.( newValue );
+										}
+									} }
+									min={ isValueUnitRelative ? 0.75 : 12 }
+									max={ isValueUnitRelative ? 6.25 : 100 }
+									step={ isValueUnitRelative ? 0.05 : 1 }
```
I chose these new `em`/`rem` values to be roughly equivalent to the previous `px` values assuming a 16px base font size.

andrewserong replied:
These values feel nice to play with in the UI! 👍

talldan replied:
@glendaviesnz also made similar changes in https://github.com/WordPress/gutenberg/pull/44959 for spacing. Font size is a bit different to spacing, so not sure if it makes sense to align these values.

andrewserong commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -201,141 +192,145 @@ const UnforwardedFontSizePicker = (
+										onChange?.(
+											hasUnits
+												? newValue
+												: parseFloat( newValue )
```
Why is `parseFloat` used here now instead of `parseInt`. Is that to allow sub-pixel values when there are no units?

noisysocks replied:
I think I was trying to make it consistent with how things work when there are units. There's probably no point changing this behaviour though... I'll put it back how it was!

andrewserong commented on packages/components/src/font-size-picker/index.tsx:
```diff
@@ -201,141 +192,145 @@ const UnforwardedFontSizePicker = (
+											onChange?.( undefined );
+										} else if ( hasUnits ) {
+											onChange?.(
+												newValue + ( valueUnit ?? 'px' )
```
Just double-checking, it looks like `valueUnit` will always be a string or undefined, right? That means we can safely use `??` because it'll never accidentally concatenate `false` or `null`?

noisysocks replied:
That's right. You can hover over `valueUnit` and TypeScript will tell you that it's `string | undefined`.

andrewserong reviewed (commented):
Nice work @noisysocks, this is looking really good! Just left a couple of questions, but no real issues from the code side of things to me 👍 

From the UI perspective, these are all mostly nits / fine-tuning notes. It looks like a gutter might need to be added to the left (and possibly right) of the slider bar so that there's a little breathing room, and for consistency with the spacing controls (https://github.com/WordPress/gutenberg/pull/44858). For example, in Storybook, the thumb lines up/overlaps the reset button slightly in the right-most position:

<img width="418" alt="image" src="https://user-images.githubusercontent.com/14988353/195241567-4b25061d-b646-483c-b2cc-27a029f6ef4d.png">

For the gutter between the UnitControl and the slider, there's a bit of discussion on @aaronrobertshaw's PR https://github.com/WordPress/gutenberg/pull/44858#issuecomment-1274173502 where @jasmussen mentioned `16px`.

Similar to that PR, I noticed that in the 100% position, the tooltip results in a scrollbar being introduced in browsers that have visible scrollbars switched on — not sure if that's also something to be addressed via gutters 🤔 (I suspect it's because the thumb hangs over the edge of the space for the component?):

<img src="https://user-images.githubusercontent.com/14988353/195241956-b96e9ce6-fca4-4030-95c3-c3e10edd49ed.png" width="300px" />

Last one: (not sure if this is something to be looked into separately), the click and drag behaviour on the UnitControl currently allows negative values to be used, so if you're scrolling down below 0, negative values are displayed and the slider pops into the middle position. Would it be worth adding a `min` value to the `<FontSizePicker>`?

<img src="https://user-images.githubusercontent.com/14988353/195242566-92bf5e3b-6f68-474e-8649-e6a2fe85e3ef.png" width="320px" />

Other than that, this looks very close to me!

ramonjd commented:
I just rebased this onto trunk to get the latest fluid type commits and tested in TT3.

I noticed that, for some custom fields, I can't change the unit:


<img src="https://user-images.githubusercontent.com/6458278/195251459-95dd636a-847e-462f-b6c2-7cc6bb8c4ec0.gif" width="250" />

Maybe it's something to do with the style value defined in theme.json. 

Here's the [h2](https://github.com/WordPress/twentytwentythree/blob/trunk/theme.json#L632), which is a clamp value. And [h4](https://github.com/WordPress/twentytwentythree/blob/trunk/theme.json#L643): a CSS var.

And [h1](https://github.com/WordPress/twentytwentythree/blob/trunk/theme.json#L626), which is a simple `value + unit`.

I can change the units for h1 but not h4 and h2.

noisysocks committed:
Add 8px margin to slider

noisysocks committed:
Add min=0 to UnitControl

noisysocks committed:
Go back to rounding pixel values when there are no units

noisysocks commented:
Thanks for the reviews! 🙇 

> It looks like a gutter might need to be added to the left (and possibly right) of the slider bar so that there's a little breathing room, and for consistency with the spacing controls (#44858).

Good flag! I added a 8px margin to the left and right of the slider. When combined with the existing 8px flex gap that makes for a 16px space between the unit control and slider. (We can't set the flex gap to 16px because then the unit control won't align with other single column controls in Global Styles.)

<img width="289" alt="Screen Shot 2022-10-13 at 14 13 46" src="https://user-images.githubusercontent.com/612155/195491083-e087d80c-80fe-463a-8d57-493e0cf1ef85.png">
<img width="337" alt="Screen Shot 2022-10-13 at 14 15 12" src="https://user-images.githubusercontent.com/612155/195491093-615ad288-c7a7-40a6-b0dc-cd4fc8888448.png">

> Last one: (not sure if this is something to be looked into separately), the click and drag behaviour on the UnitControl currently allows negative values to be used, so if you're scrolling down below 0, negative values are displayed and the slider pops into the middle position. Would it be worth adding a min value to the <FontSizePicker>?

A `min` of 0 makes total sense since negative font sizes aren't a thing. Added.

> I noticed that, for some custom fields, I can't change the unit:

@ramonjd I can't reproduce what you see there when I test in Twenty Twenty-three. Want to double check? What am I missing? 😅

ramonjd commented:
> I can't reproduce what you see there when I test in Twenty Twenty-three. Want to double check? What am I missing? 😅

Did you rebase onto trunk? I just nuked my environment, then checked out this branch and rebased and can still reproduce.

<img src="https://user-images.githubusercontent.com/6458278/195492806-d75d4750-d7dd-4985-97b7-f47eb7a33f77.gif" width="250" />

Running WordPress 6.1-RC1-54506 

 I'm very willing to accept that it's just me by the way 😄

noisysocks committed:
Restore previous logic for determining hasUnits

noisysocks commented:
Thanks @ramonjd, I did indeed forget to rebase `trunk` 😀

That bug should be fixed now. Turns out I was wrong about https://github.com/WordPress/gutenberg/pull/44598#discussion_r986406096. If we stop doing the naive check for `typeof value === 'string'` then `hasUnits` won't be true when the value is a `clamp()` value which means that the user can't select a unit. I restored the logic back to how it is in `trunk`.

Thinking about this, probably what makes more sense is to have the caller of `FontSizePicker` use an argument to say what units are supported rather than `FontSizePicker` inferring it from `value` and `fontSizes`. That's a PR for another day, though.

noisysocks commented:
Keen to get @ciampo to double check the code here but in particular my assumption that it's an "enhancement" and not a "breaking change" to change how the component looks when `withSlider` is set.

noisysocks committed:
Guard against missing first size

ramonjd commented:
>  That bug should be fixed now. 

I can confirm that it's fixed, thanks!

https://user-images.githubusercontent.com/6458278/195757227-3cbbdcd1-87d2-4460-88ae-50cdcc06007a.mp4

> Thinking about this, probably what makes more sense is to have the caller of FontSizePicker use an argument to say what units are supported rather than FontSizePicker inferring it from value and fontSizes. That's a PR for another day, though.

And have the incoming units populate the drop down? That makes sense! Agree it's a PR for "later Rob" 😄 

I'm not sure if this a UX problem or not. If I want a font size of > 100 anything, the slider becomes a little redundant in that I can adjust around my selected value, only between 0-100. And if I touch the slider my custom font size is reset to 100.


https://user-images.githubusercontent.com/6458278/195759216-3e386d0f-7e7a-4a10-8ebd-1829feac4d11.mp4

My brain expected the slider to adjust the value in the UI control.

I get that it might not be practical to have the slider adjust the range according to the custom value however, I'm not sure what's right here.

noisysocks commented:
Hm not sure what we could do about that. Dynamically set the max of the slider? It could get confusing then that different positions on the slider correspond to different values depending on what you previously entered. For what it's worth this is how the UnitControl + RangeSlider pattern works in other parts of the block editor that uses it. (Margins and padding, specifically.)

ciampo commented:
Catching up with my review queue, I'll try to give this PR a thorough look next week — but in the meantime:

- I also think this can be classified as an enhancement, given how high-level this component is (i.e. we're not making a substantial change to a primitive component)
- Re the max value of input and slider, my gut feeling is that those two components should be kept in sync when they're shown together, and should therefore have the same min and max boundaries

ramonjd commented:
> For what it's worth this is how the UnitControl + RangeSlider pattern works in other parts of the block editor that uses it. 

Ah that's interesting thanks. Maybe we could decide on a universal behaviour here and perform a sidebar-wide revisit in another PR?

ciampo commented:
> > For what it's worth this is how the UnitControl + RangeSlider pattern works in other parts of the block editor that uses it.
> 
> Ah that's interesting thanks. Maybe we could decide on a universal behaviour here and perform a sidebar-wide revisit in another PR?

This would be important to define for the work that @aaronrobertshaw is doing on `SliderControl` and the following steps ([see masterplan here](https://github.com/WordPress/gutenberg/issues/40507)), which aim at creating a high-level component that shows a number/unit control side by side with a slider.

aaronrobertshaw commented:
Thanks for the ping. 

In the `SliderNumberControl` & `SliderUnitControl` I am working on, it accepts `min`/`max` props and these are passed on both to the `NumberControl` (or `UnitControl`) and the `SliderControl` (slider portion of the RangeControl essentially). So for these components, the inner controls `max` will be kept in sync.

These components have been bumped down my priority list for a little while but I'll get something up and reviewable for them in the near future. It's also on my radar to update all occurrences in our sidebars of unit/number controls paired with range controls to these new components once they're available. This would include components such as `FontSizePicker`.


https://user-images.githubusercontent.com/60436221/196569444-d77a8044-9127-40db-aad2-020f76bba6ac.mp4

noisysocks commented:
> Re the max value of input and slider, my gut feeling is that those two components should be kept in sync when they're shown together, and should therefore have the same min and max boundaries

What would the `max` be for a font size though? Any positive number is technically valid 😅 Should we pick an arbitrary largish number e.g. 1000?

ramonjd commented:
> What would the max be for a font size though? Any positive number is technically valid 😅 Should we pick an arbitrary largish number e.g. 1000?

I'm not sure about this, but about updating the `step` and `max` attributes of the range dynamically? I mean everything over a threshold, e.g., `100`, would be the new max and `step === rangeInput.value / STEP_DEMONINATOR`

So, starting with a default max of `100` and base denominator of `10` for the step:

![2022-10-24 13 55 54](https://user-images.githubusercontent.com/6458278/197439863-c90f38ee-ae57-48a9-a3ea-1612e4f7b4b1.gif)

noisysocks committed:
Use existing parseQuantityAndUnitFromRawValue

noisysocks commented:
I do like that. But how about we implement that logic in `SliderUnitControl` and stick to the status quo for now? That way the behaviour will be consistent across components.

noisysocks committed:
Use same min, max and step values as in SpacingSizesControl

noisysocks commented:
I rebased this, removed `parseNumberAndUnit` in favour of the existing `parseQuantityAndUnitFromRawValue`, and have made the `min`, `max` and `step` values consistent with `SpacingSizesControl`. How are we feeling? 🙂

aaronrobertshaw commented:
> I do like that. But how about we implement that logic in SliderUnitControl and stick to the status quo for now? That way the behaviour will be consistent across components.

I've added this to the [tracking issue](https://github.com/WordPress/gutenberg/issues/40507) and will look to work the dynamic max/step into the `SliderUnitControl`.

ramonjd reviewed (approved):
LGTM.

This is working as described. I tested using different units and for elements, blocks in Global styles, as well as at the block-level.

Does this need a CHANGELOG entry?

Another possible follow up might be to see if adding `touch-action: none;` to the custom input numeric field improves interactivity on touch devices:

Without
![2022-10-24 15 14 29](https://user-images.githubusercontent.com/6458278/197447441-45053fdb-253a-448a-931c-e11ab3b05451.gif)



With
![2022-10-24 15 14 14](https://user-images.githubusercontent.com/6458278/197447447-0831960c-08c4-4997-992a-22d2beca6ab3.gif)

noisysocks committed:
Update CHANGELOG.md

aaronrobertshaw commented:
> Another possible follow up might be to see if adding touch-action: none; to the custom input numeric field improves interactivity on touch devices:

This was something I'd added to the new SliderNumberControl etc. Didn't see any issues in initial testing but had flagged it for a call out when it comes time for reviews.

noisysocks commented:
> Another possible follow up might be to see if adding touch-action: none; to the custom input numeric field improves interactivity on touch devices:

Looks like there's an issue for this: https://github.com/WordPress/gutenberg/issues/38865

> Does this need a CHANGELOG entry?

Thanks, added!

ciampo commented on packages/components/src/font-size-picker/test/utils.ts:
```diff
@@ -41,46 +37,6 @@ describe( 'isSimpleCssValue', () => {
-const splitValuesCases: [
-	number | string,
-	string | undefined,
-	string | undefined
-][] = [
-	// Test integers and non-integers.
-	[ 1, '1', undefined ],
-	[ 1.25, '1.25', undefined ],
-	[ '123', '123', undefined ],
-	[ '1.5', '1.5', undefined ],
-	[ '0.75', '0.75', undefined ],
-	// Valid simple CSS values.
-	[ '20px', '20', 'px' ],
-	[ '0.8em', '0.8', 'em' ],
-	[ '2rem', '2', 'rem' ],
-	[ '1.4vw', '1.4', 'vw' ],
-	[ '0.4vh', '0.4', 'vh' ],
-	// Invalid negative values,
-	[ '-5px', undefined, undefined ],
-	// Complex CSS values that shouldn't parse.
-	[ 'abs(-15px)', undefined, undefined ],
-	[ 'calc(10px + 1)', undefined, undefined ],
-	[ 'clamp(2.5rem, 4vw, 3rem)', undefined, undefined ],
-	[ 'max(4.5em, 3vh)', undefined, undefined ],
-	[ 'min(10px, 1rem)', undefined, undefined ],
-	[ 'minmax(30px, auto)', undefined, undefined ],
-	[ 'var(--wp--font-size)', undefined, undefined ],
-];
-
-describe( 'splitValueAndUnitFromSize', () => {
-	test.each( splitValuesCases )(
-		'given %p as argument, returns value = %p and unit = %p',
-		( cssValue, expectedValue, expectedUnit ) => {
-			const [ value, unit ] = splitValueAndUnitFromSize( cssValue );
-			expect( value ).toBe( expectedValue );
-			expect( unit ).toBe( expectedUnit );
-		}
-	);
-} );
-
```
Would it be useful to add these tests to the `UnitControl` utils that `FontSizePicker` is now reusing?

noisysocks replied:
Good plan. I'll spin up a PR.
